sdk/log/DESIGN.md

```diff
@@ -0,0 +1,183 @@
+}
+
+func NewLoggerProvider(...Option) *LoggerProvider
+
+func (*LoggerProvider) Logger(name string, options ...log.LoggerOption) log.Logger
+
+type Option interface { /* ... */ }
+
+func WithResource(*resource.Resource) Option
+```
+
+## LogRecord limits
+
+The [LogRecord limits](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecord-limits)
+can be configured using following options:
+
+```go
+func WithAttributeCountLimit(limit int) Option
+
+func WithAttributeValueLengthLimit(limit int) Option
+```
+
+The limits can be also configured using the `OTEL_LOGRECORD_*` environment variables as
+[defined by the specification](https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/#logrecord-limits).
+
+The options take precedence over environmental variables.
+
+### LogRecordProcessor and LogRecordExporter  
+
+Both [LogRecordProcessor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor)
+and [LogRecordExporter](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordexporter)
+are defined via an `Exporter` interface:[^2]
+
+```go
+func WithExporter(Exporter) Option {
+	return nil
+}
+
+type Exporter interface {
+	Export(ctx context.Context, records []*Record) error
+	Shutdown(ctx context.Context) error
+	ForceFlush(ctx context.Context) error
+}
+```
+
+The `Record` struct represents the [ReadWriteLogRecord](https://opentelemetry.io/docs/specs/otel/logs/sdk/#readwritelogrecord).
+
+```go
+type Record struct { /* ... */ }
+
+func (r *Record) Timestamp()
+
+func (r *Record) SetTimestamp(t time.Time)
+
+func (r *Record) ObservedTimestamp() time.Time
+
+func (r *Record) SetObservedTimestamp(t time.Time)
+
+func (r *Record) Severity() log.Severity
+
+func (r *Record) SetSeverity(level log.Severity)
+
+func (r *Record) SeverityText() string
+
+func (r *Record) SetSeverityText(text string)
+
+func (r *Record) Body() log.Value
+
+func (r *Record) SetBody(v log.Value)
+
+func (r *Record) WalkAttributes(f func(log.KeyValue) bool)
+
+// SetAttributes sets and overrides the attributes of the log record.
+func (r *Record) SetAttributes(attrs ...log.KeyValue)
+
+func (r *Record) TraceID() trace.TraceID
+
+func (r *Record) SpanID() trace.SpanID
+
+func (r *Record) TraceFlags() trace.TraceFlags
+
+func (r *Record) Resource() resource.Resource
+
+func (r *Record) InstrumentationScope() instrumentation.Scope
+
+func (r *Record) AttributeValueLengthLimit() int
+
+func (r *Record) AttributeCountLimit() int
+```
+
+The user can implement a custom [LogRecordProcessor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor)
+by implementing a `Exporter` decorator.
+
+This is similar to the design of HTTP server middleware
+which is a wrapper of `http.Handler`.[^3]
+
+[Simple processor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#simple-processor)
+is achieved by simply passing a bare-exporter.
+
+[Batching processor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#batching-processor)
```

Comment:
This differs from the way the trace export/processors work together.  Consistency seems nice for those familiar with our trace SDK.  Did you consider adopting the same model?

Reply:
See https://github.com/open-telemetry/opentelemetry-go/pull/4954#discussion_r1515838005